File: UEReflectWatch/Statestore.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace UEReflectWatch
{
    public sealed class FileMacroState
    {
        [JsonPropertyName("filePath")]
        public string FilePath { get; set; } = string.Empty;

        [JsonPropertyName("lastScanned")]
        public string LastScanned { get; set; } = string.Empty;

        [JsonPropertyName("macros")]
        public List<SerializableMacro> Macros { get; set; } = new();
    }

    public sealed class SerializableMacro
    {
        [JsonPropertyName("kind")]
        public string Kind { get; set; } = string.Empty;

        [JsonPropertyName("line")]
        public int Line { get; set; }

        [JsonPropertyName("raw")]
        public string Raw { get; set; } = string.Empty;
    }

    public sealed class StateStore
    {
        private readonly string _storePath;
        private Dictionary<string, FileMacroState> _state = new(StringComparer.OrdinalIgnoreCase);

        private static readonly JsonSerializerOptions JsonOptions = new()
        {
            WriteIndented = true
        };

        public StateStore() : this(
            Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                "UEReflectWatch"))
        {
        }

        public StateStore(string globalStoragePath)
        {
            if (!Directory.Exists(globalStoragePath))
                Directory.CreateDirectory(globalStoragePath);

            _storePath = Path.Combine(globalStoragePath, "macro-state.json");
            Load();
        }

        private void Load()
        {
            try
            {
                if (!File.Exists(_storePath)) return;
                var raw = File.ReadAllText(_storePath);
                _state = JsonSerializer.Deserialize<Dictionary<string, FileMacroState>>(raw, JsonOptions)
                         ?? new Dictionary<string, FileMacroState>(StringComparer.OrdinalIgnoreCase);
            }
            catch
            {
                _state = new Dictionary<string, FileMacroState>(StringComparer.OrdinalIgnoreCase);
            }
        }

        private void Save()
        {
            try
            {
                var json = JsonSerializer.Serialize(_state, JsonOptions);
                File.WriteAllText(_storePath, json);
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"[UEReflectWatch] Failed to save state: {ex.Message}");
            }
        }

        public List<MacroEntry> GetMacros(string filePath)
        {
            if (!_state.TryGetValue(filePath, out var entry)) return new List<MacroEntry>();

            var result = new List<MacroEntry>();
            foreach (var m in entry.Macros)
            {
                if (Enum.TryParse<MacroKind>(m.Kind, out var kind))
                    result.Add(new MacroEntry(kind, m.Line, m.Raw));
            }
            return result;
        }

        public void SetMacros(string filePath, List<MacroEntry> macros)
        {
            _state[filePath] = new FileMacroState
            {
                FilePath = filePath,
                LastScanned = DateTime.UtcNow.ToString("o"),
                Macros = macros.ConvertAll(m => new SerializableMacro
                {
                    Kind = m.Kind.ToString(),
                    Line = m.Line,
                    Raw = m.Raw
                })
            };
            Save();
        }

        public void RemoveFile(string filePath)
        {
            _state.Remove(filePath);
            Save();
        }

        public void ClearAll()
        {
            _state.Clear();
            Save();
        }

        public Dictionary<string, FileMacroState> GetFullState() =>
            new(_state, StringComparer.OrdinalIgnoreCase);

        public List<string> GetAllFiles() =>
            new List<string>(_state.Keys);
    }
}

File: RiderPlugin/src/main/kotlin/com/ueReflectWatch/store/StateStore.kt
package com.ueReflectWatch.store

import com.ueReflectWatch.scanner.MacroEntry
import com.ueReflectWatch.scanner.MacroKind
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File
import java.nio.file.Files
import java.nio.file.Paths

@Serializable
data class SerializableMacro(
    val kind: String,
    val line: Int,
    val raw: String,
    val declarationLine: String = ""
)

@Serializable
data class FileMacroState(
    val filePath: String,
    val lastScanned: String,
    val macros: List<SerializableMacro>
)

class StateStore(storageDir: String? = null) {

    private val storeFile: File
    private val state: MutableMap<String, FileMacroState> = mutableMapOf()

    private val json = Json { prettyPrint = true; ignoreUnknownKeys = true }

    init {
        val dir = storageDir ?: defaultStorageDir()
        Files.createDirectories(Paths.get(dir))
        storeFile = File(dir, "macro-state.json")
        load()
    }

    private fun defaultStorageDir(): String {
        val os = System.getProperty("os.name").lowercase()
        return when {
            os.contains("win") ->
                "${System.getenv("LOCALAPPDATA")}\\UEReflectWatch"
            os.contains("mac") ->
                "${System.getProperty("user.home")}/Library/Application Support/UEReflectWatch"
            else ->
                "${System.getProperty("user.home")}/.config/UEReflectWatch"
        }
    }

    private fun load() {
        try {
            if (!storeFile.exists()) return
            val raw = storeFile.readText()
            val loaded = json.decodeFromString<Map<String, FileMacroState>>(raw)
            state.putAll(loaded)
        } catch (_: Exception) {
            state.clear()
        }
    }

    private fun save() {
        try {
            storeFile.writeText(json.encodeToString(state))
        } catch (e: Exception) {
            System.err.println("[UEReflectWatch] Failed to save state: ${e.message}")
        }
    }

    fun getMacros(filePath: String): List<MacroEntry> {
        return state[filePath]?.macros?.mapNotNull { m ->
            runCatching { MacroEntry(MacroKind.valueOf(m.kind), m.line, m.raw, m.declarationLine) }.getOrNull()
        } ?: emptyList()
    }

    fun setMacros(filePath: String, macros: List<MacroEntry>) {
        state[filePath] = FileMacroState(
            filePath = filePath,
            lastScanned = java.time.Instant.now().toString(),
            macros = macros.map { SerializableMacro(it.kind.name, it.line, it.raw, it.declarationLine) }
        )
        save()
    }

    fun removeFile(filePath: String) {
        state.remove(filePath)
        save()
    }

    fun clearAll() {
        state.clear()
        save()
    }

    fun getAllFiles(): List<String> = state.keys.toList()

    fun getFullState(): Map<String, FileMacroState> = state.toMap()
}
